File: fc-python/FCmathparallel.py
import concurrent.futures
import json
import sympy
import math
import openai
from config import OPENAI_API_KEY, BASE_URL, MODEL_NAME
client=openai.OpenAI(
    base_url=BASE_URL,
    api_key=OPENAI_API_KEY
)
# Define all your math functions
def add(a: float, b: float):
    print("addition function called")
    return a + b
def subtract(a: float, b: float):
    print("subtraction function called")
    return a - b
def multiply(a: float, b: float):
    print("multiplication function called")
    return a * b
def divide(a: float, b: float):
    print("division function called")
    return "Error" if b == 0 else a / b
def sqrt(a: float):
    print("square root function called")
    return math.sqrt(a)
def derivative(func: str, x: float):
    print("derivative function called")
    try:
        expr = sympy.sympify(func)
        return str(sympy.diff(expr, sympy.Symbol('x')).subs(sympy.Symbol('x'), x))
    except Exception as e:
        return str(e)
def integrate(func: str, x: float):
    print("integration function called")
    try:
        expr = sympy.sympify(func)
        return str(sympy.integrate(expr, sympy.Symbol('x')).subs(sympy.Symbol('x'), x))
    except Exception as e:
        return str(e)

# Tool name to function mapping
available_functions = {
    "add": add,
    "subtract": subtract,
    "multiply": multiply,
    "divide": divide,
    "sqrt": sqrt,
    "derivative": derivative,
    "integrate": integrate,
}

# Simulate parallel function calling
def function_call_playground(prompt: str, client):
    try:
        messages = [
            {
                "role": "system",
                "content": (
                    "You are a math assistant. Given a question, always respond with a JSON array "
                    "of function calls (not explanations). Each function call should have a 'name' "
                    "and an 'arguments' field. Example:\n"
                    "[{\"name\": \"add\", \"arguments\": {\"a\": 2, \"b\": 3}},"
                    " {\"name\": \"multiply\", \"arguments\": {\"a\": 4, \"b\": 5}}]"
                )
            },
            {"role": "user", "content": prompt}
        ]

        response = client.chat.completions.create(
            model=MODEL_NAME,
            messages=messages,
            temperature=0.01,
            stream=False,
            top_p=0.95
        )

        tool_call_json = response.choices[0].message.content.strip()

        # Try parsing JSON from model output
        try:
            tool_calls = json.loads(tool_call_json)
        except json.JSONDecodeError:
            return f"Invalid JSON output: {tool_call_json}"

        # Parallel execution
        results = {}

        def call_function(tc):
            name = tc["name"]
            args = tc["arguments"]
            if name in available_functions:
                return (name, available_functions[name](**args))
            else:
                return (name, f"Unknown function '{name}'")

        with concurrent.futures.ThreadPoolExecutor() as executor:
            futures = {executor.submit(call_function, tc): tc["name"] for tc in tool_calls}
            for future in concurrent.futures.as_completed(futures):
                fname, result = future.result()
                results[fname] = result

        return results

    except Exception as e:
        return f"Error in function_call_playground: {str(e)}"

if __name__ == "__main__":
    try:
        # Test the function with a simple math problem
        prompt = "What is 2 + 3 and then multiply the result by 4?"
        print("Sending prompt:", prompt)
        result = function_call_playground(prompt, client)
        print("Result:", result)
    except Exception as e:
        print("Error in main execution:", str(e))

File: function-calling/fc-nodejs/FCmathparallel.js
import { OpenAI } from "openai";
import { config } from "dotenv";
import { BASE_URL, OPENAI_API_KEY, MODEL_NAME } from "./config.js";
import * as mathjs from "mathjs";

config();

const openai = new OpenAI({
    apiKey: OPENAI_API_KEY,
    baseURL: BASE_URL,
});

// Define all your math functions
const add = (a, b) => {
    console.log("addition function called");
    return Number(a) + Number(b);
};

const subtract = (a, b) => {
    console.log("subtraction function called");
    return Number(a) - Number(b);
};

const multiply = (a, b) => {
    console.log("multiplication function called");
    return Number(a) * Number(b);
};

const divide = (a, b) => {
    console.log("division function called");
    return Number(b) === 0 ? "Error" : Number(a) / Number(b);
};

const sqrt = (a) => {
    console.log("square root function called");
    return mathjs.sqrt(Number(a));
};

const derivative = (func, x) => {
    console.log("derivative function called");
    try {
        const expr = mathjs.parse(func);
        const symbol = mathjs.parse('x');
        const derivative = mathjs.derivative(expr, symbol);
        return derivative.evaluate({x: Number(x)}).toString();
    } catch (error) {
        return error.toString();
    }
};

const integrate = (func, x) => {
    console.log("integration function called");
    try {
        const expr = mathjs.parse(func);
        // Handle basic integration cases
        if (expr.type === 'SymbolNode' && expr.name === 'x') {
            return (Number(x) * Number(x) / 2).toString();
        } else if (expr.type === 'OperatorNode' && expr.op === '^') {
            if (expr.args[0].name === 'x') {
                const power = expr.args[1].value;
                return (Math.pow(Number(x), power + 1) / (power + 1)).toString();
            }
        }
        return "Integration not supported for this function";
    } catch (error) {
        return error.toString();
    }
};

// Tool name to function mapping
const available_functions = {
    "add": add,
    "subtract": subtract,
    "multiply": multiply,
    "divide": divide,
    "sqrt": sqrt,
    "derivative": derivative,
    "integrate": integrate,
};

// Simulate parallel function calling
export const function_call_playground_parallel = async (prompt) => {
    try {
        const messages = [
            {
                role: "system",
                content: (
                    "You are a math assistant. Given a question, always respond with a JSON array " +
                    "of function calls (not explanations). Each function call should have a 'name' " +
                    "and an 'arguments' field. Example:\n" +
                    "[{\"name\": \"add\", \"arguments\": {\"a\": 2, \"b\": 3}}," +
                    " {\"name\": \"multiply\", \"arguments\": {\"a\": 4, \"b\": 5}}]"
                )
            },
            { role: "user", content: prompt }
        ];

        const response = await openai.chat.completions.create({
            model: MODEL_NAME,
            messages: messages,
            temperature: 0.01,
            stream: false,
            top_p: 0.95
        });

        const tool_call_json = response.choices[0].message.content.trim();

        // Try parsing JSON from model output
        let tool_calls;
        try {
            tool_calls = JSON.parse(tool_call_json);
        } catch (error) {
            return { isValid: false, error: `Invalid JSON output: ${tool_call_json}` };
        }

        // Parallel execution using Promise.all
        const results = {};

        const call_function = async (tc) => {
            const name = tc.name;
            const args = tc.arguments;
            if (name in available_functions) {
                return [name, await available_functions[name](...Object.values(args))];
            } else {
                return [name, `Unknown function '${name}'`];
            }
        };

        // Execute all functions in parallel
        const promises = tool_calls.map(call_function);
        const function_results = await Promise.all(promises);

        // Build results object
        for (const [fname, result] of function_results) {
            results[fname] = result;
        }

        return {
            isValid: true,
            results: results
        };

    } catch (error) {
        return {
            isValid: false,
            error: `Error in function_call_playground: ${error.message}`
        };
    }
};

// Test the function
if (import.meta.url === `file://${process.argv[1]}`) {
    try {
        const prompt = "What is 10 + 10, then times the result by 10?";
        const start_time = Date.now();
        const result = await function_call_playground_parallel(prompt);
        const end_time = Date.now();
        console.log("Result:", result);
        console.log(`Execution time: ${(end_time - start_time) / 1000} seconds`);
    } catch (error) {
        console.error("Error in main execution:", error);
    }
} 